packages/core/lib/utils/parseUtils.ts

```diff
@@ -3,15 +3,20 @@ import type { ZodSchema } from 'zod'
 
 import type { MessageInvalidFormatError, MessageValidationError } from '../errors/Errors'
 
+export type ParseMessageResult<T> = {
+  originalMessage: T
+  parsedMessage: T
+}
+
 export function parseMessage<T extends object>(
   messagePayload: unknown,
   schema: ZodSchema<T>,
   errorProcessor: ErrorResolver,
-): Either<MessageInvalidFormatError | MessageValidationError, T> {
+): Either<MessageInvalidFormatError | MessageValidationError, ParseMessageResult<T>> {
   try {
-    return {
-      result: schema.parse(messagePayload),
-    }
+    const parsedMessage = schema.parse(messagePayload)
+
+    return { result: { parsedMessage, originalMessage: messagePayload as T } }
```

Comment:
might be helpful to add comment, explaining why we have both

Reply:
Same as above